### PR TITLE
[DTensor][1/N] add forward layer norm support

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113244
* __->__ #113105

**Summary**:
This PR adds DTensor implementation for ATen op `native_layer_norm`.

**Test**:
`pytest test/distributed/_tensor/test_dtensor_ops.py -s -k layer_norm`

cc @wanchaol

### CODE REVIEW

File: torch/distributed/_tensor/ops/math_ops.py [hunk 1]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+            )
+        )
+
+    return output_strategy
+
+
+def _replicate_dims_start_at(
+    placements: Sequence[Placement], starting_dim_optional: Optional[int] = 0
+) -> Tuple[Placement, ...]:
+    starting_dim = 0 if starting_dim_optional is None else starting_dim_optional
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:338]: if you are doing this check, why not just make `starting_dim_optional: int = 0`?

File: torch/distributed/_tensor/ops/math_ops.py [hunk 2]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+    new_placements = []
+    for idx, placement in enumerate(placements):
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:340]: I think this looks wrong, you are misusing `mesh` idx and the tensor dim idx. 

What we want is to make sure the tensor dimension starting from a certain dimension, there's no sharding, but it seems that here we are checking the mesh idx instead.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 3]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+    # must be in form of OpStrategy
+    assert isinstance(input_schema, OpStrategy)
+    assert isinstance(normalized_shape, (int, Sequence, torch.Size))
+    normalized_size = (
+        [normalized_shape]
+        if isinstance(normalized_shape, int)
+        else list(normalized_shape)
+    )
+
+    input_shape = input_schema.output_shape
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:276]: this is not used?

File: torch/distributed/_tensor/ops/math_ops.py [hunk 4]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+        redistribute_costs.append(
+            generate_redistribute_costs(input_schema, input_target_spec)
+        )
+
+        if weight_schema is not None:
+            assert isinstance(weight_schema, OpStrategy)
+            weight_src_spec = weight_schema.strategies[idx].output_spec
+            weight_target_spec = DTensorSpec(
+                mesh=mesh,
+                placements=_replicate_dims_start_at(weight_src_spec.placements, 0),
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:303]: are we saying that we only support weight/bias replication for now? worth adding a comment

**Review by wanchaol — COMMENTED** [review]
first pass, please see the inline comments, I think the semantics of `replicate_dim_start_at` looks wrong and it might make all the tensor dims be replicated.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 5]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
     ):
         raise RuntimeError("Cannot run _softmax_backward_data on sharding dimension!")
     return pointwise_rule(op_schema)
+
+
+def layer_norm_strategy(mesh: DeviceMesh, op_schema: OpSchema) -> OpStrategy:
+    # args must be: input, normalized_shape, weight, bias, eps
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:257]: Good question. I did two things to testify:
1. `assert len(op_schema.args_schema) == 5` on line 258 to see if the schema is always a quintuple and the dtensor test has no error raised. 
2. However, this could be likely due to we always passed in `weight` and `bias` so I also added `assert bias is not None`. And the dtensor test raises error saying `bias` could be `None`. So we can tell that dtensor test actually tests the case where `bias` is optional.

Based on these two evidence, `weight`/`bias` always exist even if they're `None`.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 6]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+    assert len(op_schema.args_schema) == 5
+    (
+        input_schema,
+        normalized_shape,
+        weight_schema,
+        bias_schema,
+        _,
+    ) = op_schema.args_schema
+    # the current layer norm implementation requires that all input DTensor's sharding
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:266]: Are strategy based op implementation mixed with ShardingProp based op implementation in sharding propagation? If so, I'll need to add compatible code to handle that.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 7]
```diff
@@ -251,3 +251,100 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+    # must be in form of OpStrategy
+    assert isinstance(input_schema, OpStrategy)
+    assert isinstance(normalized_shape, (int, Sequence, torch.Size))
+    normalized_size = (
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:270]: oops, the code was override in rebase...

File: torch/distributed/_tensor/ops/math_ops.py (hunk 1 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:338]: that's right. I keep mixing default arguments with optional type hints, in a wrong way....

File: torch/distributed/_tensor/ops/math_ops.py (hunk 2 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:340]: i see. that's why the code in other places is using `dim_map`....

File: torch/distributed/_tensor/ops/math_ops.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/ops/math_ops.py:266]: they are handled separately now, eventually we will get rid of the sharding prop rule based implementation.

File: torch/distributed/_tensor/ops/math_ops.py (hunk 4 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:303]: Yes, to be more specific, we allow sharded weight/bias but we will replicate them before feeding them into the op.

File: torch/distributed/_tensor/ops/math_ops.py [hunk 8]
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+    new_placements: List[Placement] = []
+    for p in placements:
+        if p.is_partial() or (isinstance(p, Shard) and p.dim < start_dim):
+            new_placements.append(Replicate())  # make it replicate
+        else:
+            new_placements.append(p)  # keep the placement
+    return tuple(new_placements)
+
+
+register_op_strategy(aten.native_layer_norm.default, schema_info=RuntimeSchemaInfo())(
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:355]: why not use make this a decorator for the layernorm strategy? It's the only one that uses this strategy

File: torch/distributed/_tensor/ops/math_ops.py [hunk 9]
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+
+    return output_strategy
+
+
+def _replicate_dims_start_at(
+    placements: Sequence[Placement], start_dim: int = 0
+) -> Tuple[Placement, ...]:
+    new_placements: List[Placement] = []
+    for p in placements:
+        if p.is_partial() or (isinstance(p, Shard) and p.dim < start_dim):
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:348]: hmmm shouldn't this be `p.dim >= start_dim` instead?

File: torch/distributed/_tensor/ops/math_ops.py [hunk 10]
```diff
@@ -252,3 +253,105 @@ def softmax_bwd_rule(op_schema: OpSchema) -> OutputSharding:
...
+        op_args_target_specs.append(input_target_spec)
+        redistribute_costs.append(
+            generate_redistribute_costs(input_strategy, input_target_spec)
+        )
+
+        if weight_strategy is not None:
+            assert isinstance(weight_strategy, OpStrategy)
+            weight_src_spec = weight_strategy.strategies[idx].output_spec
+
+            # for the weight tensor, we replicate it on all dims if necessary
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:304]: can you add a TODO here so that we remember to see if we can keep weight/bias sharded instead of replicating on all dims?

File: torch/distributed/_tensor/ops/math_ops.py [hunk 11]
```diff
@@ -1,5 +1,5 @@
 # Copyright (c) Meta Platforms, Inc. and affiliates
-from typing import cast, List, Optional, Tuple
+from typing import cast, List, Optional, Sequence, Tuple
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/math_ops.py:2]: In addition to the op level tests, I think we should add a test in `test_math_ops.py` to use `distribute_module` to partition a nn.LayerNorm, and mark input as sharded in `prepare_input_fn`, to make sure things work for the forward case when we apply this to nn.Module, this can be a follow up PR

**Review by wanchaol — COMMENTED** [review]
Have a few comments inlined otherwise good to go!

File: torch/distributed/_tensor/ops/math_ops.py (hunk 10 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py:304]: quick question: in short, what is the ideal way to support sharded weight/bias? My understanding is op decompose. Is that right?

Comment (wanchaol) [reply torch/distributed/_tensor/ops/math_ops.py:310]: I think we almost likely won't need to make things sharded there, if we are doing it the normalized shape also need to allow sharding, we probably need some sort of decomposition there. Right now let's just have the complete picture documented.

**Review by wanchaol — APPROVED** [review]
lgtm!

File: torch/distributed/_tensor/ops/math_ops.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/ops/math_ops.py]: note: forward fix `normalize_to_torch_size` in #113244 to handle the `size` argument of type `int`, `Sequence`, or `torch.Size`. This fix also enables 3 dtensor op tests.
